parser/scanner.go

```diff
@@ -8,38 +8,41 @@ import (
 
 type Scanner struct {
 	src         source // the source the scanner is drawing from
-	sliceStart  int    // the start of the slice visible to the scanner
-	sliceLength int    // the length of the slice visible to the scanner
+	sliceStart  int    // the start of the slice visible to the scanner, based on the original src
+	sliceLength int    // the length of the slice visible to the scanner, based on the original src
 }
 
 type source interface {
 	length() int                // the length of the entire source string
 	slice(i, length int) string // the string of the given slice
 	filename() string           // the name of the file from which the source is derived (or empty if none)
+	stripSource(i, length int) source
 }
 
 type stringSource struct {
-	origin *string // the entire source string
-	f      string  // the source filename
+	origin string // the entire source string
+	f      string // the source filename
 }
 
 func NewScanner(str string) *Scanner {
-	return &Scanner{stringSource{origin: &str}, 0, len(str)}
+	return &Scanner{stringSource{origin: str}, 0, len(str)}
 }
 
 func NewScannerWithFilename(str, filename string) *Scanner {
-	return &Scanner{stringSource{&str, filename}, 0, len(str)}
+	return &Scanner{stringSource{str, filename}, 0, len(str)}
 }
 
 func NewScannerAt(str string, offset, size int) *Scanner {
-	if offset+size > len(str) {
-		panic(fmt.Errorf("%d is out of range [:%d]", offset+size, len(str)))
-	}
-	return &Scanner{stringSource{origin: &str}, offset, size}
+	return &Scanner{stringSource{origin: str}, offset, size}
```

Comment:
Hmm. Removing the test is necessary, but only because arr.ai is calling it incorrectly. We should reinstate the test, but it can wait till we fix the problem end-to-end. This will probably require introduction of a `source.offset` field so `str` can represent a snippet of the original source, not the entirety.